jib-core/src/main/java/com/google/cloud/tools/jib/cache/DefaultCacheStorageFiles.java

```diff
@@ -113,6 +116,33 @@ Path getLayerDirectory(DescriptorDigest layerDigest) {
     return getLayersDirectory().resolve(layerDigest.getHash());
   }
 
+  /**
+   * Gets the directory to store the image manifest and configuration.
+   *
+   * @return the directory for the image manifest and configuration
+   */
+  Path getImagesDirectory() {
+    return cacheDirectory.resolve(IMAGES_DIRECTORY);
+  }
+
+  /**
+   * Gets the directory corresponding to the given image reference.
+   *
+   * @param imageReference the image reference
+   * @return a path in the form of {@code
+   *     (jib-cache)/images/registry[/port]/repository/(tag|digest-type/digest)}
+   */
+  Path getImageDirectory(ImageReference imageReference) {
+    // Split image reference on '/', ':', and '@' to build directory structure
```

Comment:
This means `my-repo:3000` and `my-repo/3000` will point to the same directory. Some random ideas:

Use some kind of mapping function (one-way hashing, encoding, or whatever) from a char to another that can be safely used as a filename on all filesystems. We may apply the mapping function to the entire image reference or only some portions (like some special characters which may or may not include `/`). But still, probably it would be good to have directory structures.

Reply:
Maybe we can just use `_` in place of of `:` in the registry name. `_` is an invalid character in the registry component of an image reference, but a valid path character.

Reply:
Unfortunately, I could use `_` with `docker tag` on both the registry component part and the tag part.

Reply:
Ok, I was misremembering then.

Reply:
Maybe we can use `!` then.

Reply:
I like the simple approach. In sum, converting `:` and `@` to `!`, right? @GoogleContainerTools/java-tools-build any objections?

Reply:
Are `@` and `:` only allowed in specific sections?
- `:` only for port
- `@` only for digest
?

are `.` allowed in the image reference?

Reply:
I would prefer `.` to `!`, exclamation looks a little crazy in filenames

Reply:
`:` is used for port, (`registry:port/...`), tag (`registry/repo:tag`), and digest (`registry/repo@sha256:aaaaa...`). And `.` is a valid character to use anywhere in the image reference, so I don't think we should use that.

Reply:
`!` does look weird, but I don't think it matters whether or not we make the paths pretty. It just needs to not be error-prone and produce a unique conversion from image reference to path.